### PR TITLE
Tiered tactic assignment

### LINKED ISSUE #2005: Change stp and update plays to support tiered tactic assigment
### Description of the task
We don't always want to globally minimize the cost of assigning all tactics to all robots. 
Sometimes we want a more greedy assignment for a smaller set of "higher priority" tactics (tactics that involve the ball, cherry pickers, etc...)

We should setup plays to yield a vector of vectors, and the cost of assigning tactics should be minimized for each of the tactics in the inner vectors, in the order of priority defined by the outer vector.

~~~
outer_vector = 
{ 
  {some_tactic}, // inner vector
  {crease_defender_1, crease_defender_2}, 
  {move_tactic},
}                   
~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

The cost of assigning a some_tactic should be minimized across all robots first,  followed by both the crease_defender tactics. The move tactic should be assigned last

<!--
    What does this work depend on?
    What interface will this work use or create?
    What are the main components of the task?
    Where does this work fit in the larger project?

    It is important to define this task sufficiently so that an untrained
    team member can take it on and know where to start. Feel free to
    link to resources or other team member which could guide the assignee to
    complete the task
-->

### Acceptance criteria

<!--
    Checkbox list that outlines what needs to be done in order for this task
    to be considered "complete".

    Specify any implementation requirements such as data structures,
    functionalities, testing requirements, documentation, etc.
-->

- [ ] Implement feature
- [ ] Add tests
- [ ] Make sure nothing breaks/no regressions

### Blocked By

<!--
    List all other issues that need to be completed before this one, ex:
    - #123
    - #374
-->

N/A

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description
We don't always want to globally minimize the cost of assigning all tactics to all robots.
Sometimes we want a more greedy assignment for a smaller set of "higher priority" tactics (tactics that involve the ball, cherry pickers, etc...)

We should setup plays to yield a vector of vectors, and the cost of assigning tactics should be minimized for each of the tactics in the inner vectors, in the order of priority defined by the outer vector.

See ticket for more details

### Testing Done
unit tests, manual testing

### Resolved Issues
resolves #2005 

### Length Justification

N/A

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — COMMENTED** [review]
looks good, just replace std::vector<std::shared_ptr<Tactic>> with TacticVector and similarly for the other aliases

**Review by LiCody — COMMENTED** [review]
lgtm , 1 small nit

File: src/software/ai/hl/stp/stp.h [hunk 1]
```diff
@@ -122,22 +124,36 @@ class STP : public HL
     PlayInfo getPlayInfo() override;
 
     /**
-     * Given a list of tactics and the current World, assigns robots from the friendly
-     * team to each tactic
+     * Given a vector of a vector of tactics and the current World, assigns robots
```
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/stp.h]: fixed

File: src/software/ai/hl/stp/stp_tactic_assignment_test.cpp [hunk 2]
```diff
@@ -485,10 +485,10 @@ TEST_F(STPTacticAssignmentTest,
 
     auto goalie_tactic_1 = std::make_shared<GoalieTestTactic>();
     auto goalie_tactic_2 = std::make_shared<GoalieTestTactic>();
-    std::vector<std::shared_ptr<const Tactic>> tactics = {goalie_tactic_1,
-                                                          goalie_tactic_2};
+    std::vector<std::vector<std::shared_ptr<const Tactic>>> tactics = {
```
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/stp_tactic_assignment_test.cpp]: fixed all use cases in this test

File: src/software/ai/hl/stp/stp.h [hunk 3]
```diff
@@ -122,22 +124,36 @@ class STP : public HL
...
+     *
+     * The order of the given tactics in the inner vector also determines their priority,
+     * with the tactics at the beginning of the vector being a higher priority than those
+     * at the end. The priority determines which tactics will NOT be assigned if there are
+     * not enough robots on the field to assign them all. For example, if a Play returned
+     * 4 Tactics in total but there were only 3 robots on the field at the time, only the
+     * first 3 Tactics in the vectors would be assigned to robots and run. (In the example
+     * above, only the goalie and crease_defenders would be assigned)
+     *
+     * @param tactics The list of list of tactics that should be assigned a robot. Note
```
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/stp.h]: fixed

**Review by jonathanlew — APPROVED** [review]

**Review by LiCody — APPROVED** [review]

**Pull request merged**